include/aspect/boundary_temperature/dynamic_core.h

```diff
@@ -0,0 +1,521 @@
+     * and the core-mantle boundaries (CMB) temperature is calculated by core energy balance.
+     * The formulation of core energy balance are from:
+     * Nimmo et al., The influence of potassium on core and geodynamo evolution.
+     *    Geophysical Journal International, 2004. 156(2): p. 363-376.
+     * @ingroup BoundaryTemperatures
+     */
+    template <int dim>
+    class DynamicCore : public Interface<dim>, public aspect::SimulatorAccess<dim>
+    {
+      public:
+        /**
+         * Construstor
+         */
+        DynamicCore();
+
+        /**
+         * Return the temperature that is to hold at a particular location on the
+         * boundary of the domain. This function returns the temperatures
+         * at the inner and outer boundaries.
+         *
+         * @param boundary_indicator The boundary indicator of the part of the boundary
+         *   of the domain on which the point is located at which we are requesting the
+         *   temperature.
+         * @param location The location of the point at which we ask for the temperature.
+         **/
+        virtual
+        double  boundary_temperature (const types::boundary_id            boundary_indicator,
+                                      const Point<dim>                    &location) const;
+
+        /**
+         * Return the minimal temperature on that part of the boundary
+         * on which Dirichlet conditions are posed.
+         *
+         * This value is used in computing dimensionless numbers such as the
+         * Nusselt number indicating heat flux.
+         */
+        virtual
+        double minimal_temperature (const std::set<types::boundary_id> &fixed_boundary_ids) const;
+
+        /**
+         * Return the maximal temperature on that part of the boundary
+         * on which Dirichlet conditions are posed.
+         *
+         * This value is used in computing dimensionless numbers such as the
+         * Nusselt number indicating heat flux.
+         */
+        virtual
+        double maximal_temperature (const std::set<types::boundary_id> &fixed_boundary_ids) const;
+
+        /**
+         * Declare the parameters this class takes through input files.
+         * This class declares the inner and outer boundary temperatures.
+         */
+        static
+        void
+        declare_parameters (ParameterHandler &prm);
+
+        /**
+         * Read the parameters this class declares from the parameter
+         * file.
+         */
+        virtual
+        void
+        parse_parameters (ParameterHandler &prm);
+
+        /**
+         * This function update the core-mantle boundary (CMB) temperature by
+         * the core energy balance solver using the core-mantle boundary heat flux.
+         */
+        virtual void update();
+
+        /**
+         * Pass core data to other modules
+         */
+        struct internal::CoreData get_core_data() const;
```

Comment:
Right. I didn't mean that one class or the other stores a reference, just that the functions that return these objects return them by const reference. In other words, you can still do
```
  this->core_data = other_object.get_core_data();
```
and the data will be *copied* from `other_object` to `this`. It's just that the way you write the code right now, it will be copied *twice*: once from `other_object.core_data` to the object that `get_core_data()` returns, and then a second time from that object to `this->core_data`.